spectator-ext-stackdriver/src/main/java/com/netflix/spectator/stackdriver/StackdriverWriter.java

```diff
@@ -0,0 +1,760 @@
+   * this only happens the first time we wish to use it over the lifetime of
+   * the project. It is also possible a extra times in other processes
+   * encountering a race condition on the first-time use where these will fail.
+   */
+  MetricDescriptor createDescriptorHelper(Id id, Meter meter)
+       throws IOException {
+    String descriptorName = idToDescriptorName(id);
+    String descriptorType = idToDescriptorType(id);
+    MetricDescriptor descriptor = new MetricDescriptor();
+    descriptor.setName(descriptorName);
+    descriptor.setType(descriptorType);
+    descriptor.setValueType("DOUBLE");
+    descriptor.setMetricKind(meterToKind(meter));
+
+    List<LabelDescriptor> labels = new ArrayList<LabelDescriptor>();
+    LabelDescriptor labelDescriptor = new LabelDescriptor();
+    labelDescriptor.setKey(APPLICATION_LABEL);
+    labelDescriptor.setValueType("STRING");
+    labels.add(labelDescriptor);
+    labelDescriptor = new LabelDescriptor();
+    labelDescriptor.setKey(INSTANCE_LABEL);
+    labelDescriptor.setValueType("STRING");
+    labels.add(labelDescriptor);
+
+    // All measurements have the same tags, so just look at the first.
+    Measurement measurement = meter.measure().iterator().next();
+    for (Tag tag : measurement.id().tags()) {
+       labelDescriptor = new LabelDescriptor();
+       labelDescriptor.setKey(tag.key());
+       labelDescriptor.setValueType("STRING");
+       labels.add(labelDescriptor);
+    }
+
+    maybeAddLabelHints(descriptorType, labels);
+    descriptor.setLabels(labels);
+    MetricDescriptor response = service.projects().metricDescriptors()
+      .create(projectResourceName, descriptor)
+      .execute();
+
+    log.info("Created new MetricDescriptor {}", response.toString());
+    return response;
+  }
+
+  /**
+   * Ensure a descriptor for the meter exists, creating one if needed.
+   */
+  public MetricDescriptor ensureDescriptor(Id id, Meter meter) {
+    if (knownDescriptors.isEmpty()) {
+      try {
+          initKnownDescriptors();
+      } catch (HttpResponseException rex) {
+        log.error(
+            "Caught HttpResponseException initializing KnownDescriptors", rex);
+        return null;
+      } catch (IOException ioex) {
+        log.error("Caught IOException initializing knownDescriptors.", ioex);
+        return null;
+      }
+    }
+
+    String descriptorType = idToDescriptorType(id);
+    MetricDescriptor found = knownDescriptors.get(descriptorType);
+    if (found != null) {
+      return found;
+    } else if (badDescriptorTypes.contains(descriptorType)) {
+      return null;
+    }
+    Iterator<Measurement> iterator = meter.measure().iterator();
+    if (!iterator.hasNext()
+              || !iterator.next().id().tags().iterator().hasNext()) {
+      // Ignore metrics whose measures do not have labels.
+      // These are gleaned from spring and the interesting
+      // ones should be replicated by other explicit Spectator
+      // metrics having labels.
```

Comment:
I wonder if this maybe overly broad. I'll think about it some more to see if I can come up with some counter examples.